clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -830,96 +872,110 @@ invariant_checker_t::print_results()
 
 std::string
 invariant_checker_t::check_for_pc_discontinuity(
-    per_shard_t *shard, const memref_t &memref,
-    const std::unique_ptr<instr_autoclean_t> &cur_instr_decoded,
-    const bool expect_encoding)
+    per_shard_t *shard, const memref_t &memref, const memref_t &prev_instr, addr_t cur_pc,
+    const std::unique_ptr<instr_autoclean_t> &cur_instr_decoded, bool expect_encoding,
+    bool at_kernel_event)
 {
     std::string error_msg = "";
-    bool have_cond_branch_target = false;
-    addr_t cond_branch_target = 0;
-    const addr_t prev_instr_trace_pc = shard->prev_instr_.instr.addr;
-    if (prev_instr_trace_pc != 0 /*first*/ &&
-        // We do not bother to support legacy traces without encodings.
-        expect_encoding && type_is_instr_direct_branch(shard->prev_instr_.instr.type)) {
-        if (shard->prev_instr_.instr.encoding_is_new)
+    bool have_branch_target = false;
+    addr_t branch_target = 0;
+    const addr_t prev_instr_trace_pc = prev_instr.instr.addr;
+    if (prev_instr_trace_pc == 0 /*first*/)
+        return "";
+    // We do not bother to support legacy traces without encodings.
+    if (expect_encoding && type_is_instr_direct_branch(prev_instr.instr.type)) {
+        if (prev_instr.instr.encoding_is_new)
             shard->branch_target_cache.erase(prev_instr_trace_pc);
         auto cached = shard->branch_target_cache.find(prev_instr_trace_pc);
         if (cached != shard->branch_target_cache.end()) {
-            have_cond_branch_target = true;
-            cond_branch_target = cached->second;
+            have_branch_target = true;
+            branch_target = cached->second;
         } else {
+            // TODO i#5912: This prev_instr_decoded_ isn't always the same as the
```

Comment:
Thanks for noting.

Question about the management of branch_target_cache: when we exit a signal, we might miss adding the target of the last instr of the signal (because here prev_instr is actually the last instr before the signal). Maybe this is not a real issue because the last instr in a signal would be sigreturn which is not a branch. But if we were tracking something else per-decoded-instr (e.g. the decoded instr_t itself), then we would want to do the updates for the actual last instr, not the last instr in the current context.

Reply:
I think, as noted, we should just get rid of branch_target_cache and use per-context direct-stored info on the per-context prior instr.